vignettes/philosophy.Rmd

```diff
@@ -0,0 +1,163 @@
+Doing so enforces you to consistenly use the same coding style, making it easy to learn and use it.
+
+### Rules for coding style
+
+- `underscore_separated` names for functions, parameters and variables.
+- A line of code or comments must be no longer than 80 characters.
+  Pro tip: have RStudio display this margin in the editor.
+  _Tools_ > _Global options_ > _Code_ > _Display_ > _Show margin_
+- Object names must not be longer than 30 characters.
+- Start a new line
+    - after the pipe (`%>%`)
+    - never before but always after `{`
+    - before `}`
+- Use spaces instead of tabs.
+  Pro tip: make RStudio place 2 spaces when you hit the tab key.
+  _Tools_ > _Global options_ > _Code_ > _Editing_ > _Insert spaces for tabs_
+- Use spaces consistently
+    - Use exactly one space before and after
+        - assignments `<-`, `->`, `=`
+        - operators like `+`, `-`,`*`, `/`, ...
+    - No space before and one space after `,`
+    - One space before `(` except in `function(` (don't use `function (`)
+    - One space between `)` and `{`
+    - No space after or before `(` or `[`
+- Use double quotes (`"`) to define character strings.
+- No trailing whitespace
+    - spaces at the end of a line
+    - blank lines at the end of the script
+
+### Static code analysis checks
+
+- Is an object defined before you use it?
+- Do you use an object after you defined it?
+- Use `<-` or `->` to assign something.
+  Only use `=` to pass arguments to a function (e.g. `check_package(fail = TRUE)`).
+- Use `is.na(x)` instead of `x == NA`.
+- Use `seq_len()` or `seq_along()` instead of `1:length(x)`, `1:nrow(x)`, ...
+  When `length(x) == 0`, `1:length(x)` yields `c(1, 0)`, whereas `seq_along(x)` would yield an empty vector.
+- Don't store code in comments.
+  If you don't want to loose code, use version control systems like [`git`](https://git-scm.com/).
+- Avoid code with lots of nested loop or if statements.
+  If the code it too complex, you'll get a warning that the [cyclomatic complexity](https://en.wikipedia.org/wiki/Cyclomatic_complexity) is too high.
+  Tips for reducing the code complexity:
+    - Use `assertthat::assert_that()` to validate object or conditions instead of `if() stop()`.
+    - See if you can use `ifelse()` instead of `if()`.
+    - Split the main function of your code over a few sub functions.
+
+## File name conventions
+
+To make this easier to remember we choose the same name conventions for file names as for objects.
+We acknowledge that these rules sometimes clash with requirements from other sources (e.g. `DESCRIPTION` in an R package, `README.md` on GitHub, `.gitignore` for git, ...).
+In such case we allow the file names as required by R, git or GitHub.
+When `check_filename()` does not allow a certain file or folder name and you think this is a false positive, then please open an [issue on GitHub](https://github.com/inbo/checklist/issues) and motivate why this is a false positive.
```

Comment:
`false positive` is a strange wording for something that gives an error, I think.  ;-)  I made a suggestion, I don't know if this is better?
```suggestion
When `check_filename()` does unfairly not allow a certain file or folder name, then please open an [issue on GitHub](https://github.com/inbo/checklist/issues) and motivate why this should be allowed.
```